File: merge_sort.c
#include <stdio.h>

int merge_sort(int nums[], int nums_length);
int merge_halfs(int nums[],int la[], int la_len, int ra[], int ra_len);

int main(void) {
    int nums[] = {-8,-31,0,7,5,-31,43};
    int nums_length = 7;
    merge_sort(nums,nums_length);
    for (int i = 0; i < nums_length; i++) {
        printf("nums element %i: %i\n", i+1,nums[i]);
    }
}

int merge_sort(int nums[], int nums_length) {
    if (nums_length == 1) 
        return 0;
    int mid = nums_length/2;
    int la[mid];
    int ra[nums_length-mid];
    for (int i = 0; i < mid; i++) {
        la[i] = nums[i];
    }
    for (int i = 0; i < nums_length-mid; i++) {
        ra[i] = nums[mid+i];
    }
    merge_sort(la, mid);
    merge_sort(ra,nums_length-mid);
    merge_halfs(nums,la,mid,ra, nums_length-mid);
}

int merge_halfs(int nums[],int la[], int la_len, int ra[], int ra_len) {
    int merged_array[la_len + ra_len];
    int la_i = 0;
    int ra_i = 0; // these are the left, right, and merged arrays respectively
    int ma_i = 0;
    
    while (la_i < la_len && ra_i < ra_len) {
        if (la[la_i] < ra[ra_i]) {
            merged_array[ma_i] = la[la_i];
            la_i++;
        } else {
            merged_array[ma_i] = ra[ra_i];
            ra_i++;
        }
        ma_i++;
    }
    while (la_i < la_len) {
        merged_array[ma_i] = la[la_i];
        la_i++;
        ma_i++;
    }
    while (ra_i < ra_len) {
        merged_array[ma_i] = ra[ra_i];
        ra_i++;
        ma_i++;
    }
    for (int i = 0; i < (la_len + ra_len); i++) {
        nums[i] = merged_array[i];
        //printf("Merged array element %i: %i\n", i+1,merged_array[i]);
    }
}

File: mergeSort.js
function mergeHalves(la,ra) {
  let mergedArray = []
  let [ laI, raI, maI ] = [0,0,0]

  while (laI < la.length && raI < ra.length) {
    if (la[laI] < ra[raI]) {
      mergedArray[maI] = la[laI]
      laI++
    } else {
      mergedArray[maI] = ra[raI]
      raI++
    }
    maI++
  }

  while (laI < la.length) {
    mergedArray[maI] = la[laI]
    laI++
    maI++
  }
  while (raI < ra.length) {
    mergedArray[maI] = ra[raI]
    raI++
    maI++
  }
  return mergedArray
}

function mergeSort(array) {
  const len = array.length
  if (len <= 1) return
  const mid = Math.floor(len/2)
  const la = []
  const ra = []

  for (let i = 0; i < mid; i++) {
    la[i] = array[i]
  }
  for (let i = 0; i < len - mid; i++) {
    ra[i] = array[i+mid]
  }

  mergeSort(la)
  mergeSort(ra)
  const mergedArray = mergeHalves(la,ra)

  for (let i = 0; i < len; i++) {
    array[i] = mergedArray[i]
  }
}

let array = [-3,-22,5,1,0,9,17,4]

mergeSort(array)

console.log(array)
